### ISSUE TITLE
TextField fails to accept newline in Linux client

### ISSUE THREAD

**Issue by JanezStupar** [flutter/flutter#62207]
## Steps to Reproduce

Given simple app:
```
import 'package:flutter/material.dart';

void main() {
  runApp(MaterialApp(
    home: new MyApp(),
  ));
}

class MyApp extends StatelessWidget {

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(title: Center(child: Text('input test'))),
      body: Center(
        child: TextField(
        keyboardType: TextInputType.multiline,
        textInputAction: TextInputAction.newline,
        expands: true,
        minLines: null,
        maxLines: null,
      ))
    );
  }
}
```
Input some text and press enter to go to next line.

**Expected result:** "Cursor should move to a new line"
**Actual result:** "Pressing enter key doesn't do anything, neither does shift + enter."
**Note:** Works fine on Mac!
<details>

```
[✓] Flutter (Channel master, 1.21.0-6.0.pre.17, on Linux, locale en_US.UTF-8)
    • Flutter version 1.21.0-6.0.pre.17 at /home/janez/PycharmProjects/tmp/flutter/flutter
    • Framework revision e64fe11542 (16 hours ago), 2020-07-23 18:11:17 -0700
    • Engine revision 626244a72c
    • Dart version 2.9.0 (build 2.9.0-21.0.dev a3815b6590)

[✗] Android toolchain - develop for Android devices
    ✗ Unable to locate Android SDK.
      Install Android Studio from: https://developer.android.com/studio/index.html
      On first launch it will assist you in installing the Android SDK components.
      (or visit https://flutter.dev/docs/get-started/install/linux#android-setup for detailed instructions).
      If the Android SDK has been installed to a custom location, set ANDROID_SDK_ROOT to that location.
      You may also want to add it to your PATH environment variable.


[✓] Linux toolchain - develop for Linux desktop
    • clang version 10.0.0
    • cmake version 3.17.3
    • ninja version 1.10.0
    • pkg-config version 1.7.3

[!] Android Studio (not installed)
    • Android Studio not found; download from https://developer.android.com/studio/index.html
      (or visit https://flutter.dev/docs/get-started/install/linux#android-setup for detailed instructions).

[✓] IntelliJ IDEA Ultimate Edition (version 2019.3)
    • IntelliJ at /opt/intellij-idea-ultimate-edition
    • Flutter plugin version 45.1.2
    • Dart plugin version 193.6911.31

[✓] Connected device (1 available)
    • Linux (desktop) • linux • linux-x64 • Linux

! Doctor found issues in 2 categories.
```

</details>

**Comment by wisedevman** [comment]
I confirm.

**Comment by maks** [comment]
@gspencergoog I had a look at editable.dart but it doesn't look like the right place, if you could point me at the right part of the codebase, I'd be keen to have a go at fixing this as its blocking a project I'm working on at the moment,

**Comment by stuartmorgan** [comment]
>  if you could point me at the right part of the codebase

See [here](https://github.com/flutter/engine/blob/master/shell/platform/linux/fl_text_input_plugin.cc#L325); compare with [macOS](https://github.com/flutter/engine/blob/master/shell/platform/darwin/macos/framework/Source/FlutterTextInputPlugin.mm#L244)

**Comment by maks** [comment]
Brillant, thanks @stuartmorgan !! I'm looking into it now.

**Comment by maks** [comment]
Thanks again for the pointers @stuartmorgan ! I think I've figured out whats happened.

Seems that in the move from GLFW to GTK, the new GTK text plugin lost the functionality to keep track of if the input type is "MultilineInputType", which all the other platform plugins do use.

So what's now missing in the GTK plugin is something like this that the GLFW plugin did:
```
if (input_type_ == kMultilineInputType) {
    model->AddCodePoint('\n');
```

and thats what the macOS one does that [you pointed me to](https://github.com/flutter/engine/blob/master/shell/platform/darwin/macos/framework/Source/FlutterTextInputPlugin.mm#L244):
```
if ([self.activeModel.inputType isEqualToString:kMultilineInputType]) {
      [self insertText:@"\n" replacementRange:self.activeModel.selectedRange];
```

So I think the fix is just to do as the glfw plugin did: add the same check for current input type and insert the newline char using the TextInputModel's `AddCodePoint()`, if the Flutter tells us that the input type is currently "multiline".

But this where I think I'd need some guidance, as in the old glfw plugin, it got input type using:
```
input_type_ = "";
    auto input_type_info_json = client_config.FindMember(kTextInputType);
    if (input_type_info_json != client_config.MemberEnd() &&
        input_type_info_json->value.IsObject()) {
      auto input_type_json =
          input_type_info_json->value.FindMember(kTextInputTypeName);
      if (input_type_json != input_type_info_json->value.MemberEnd() &&
          input_type_json->value.IsString()) {
        input_type_ = input_type_json->value.GetString();
      }
    }
```
which I assume is json config coming out of the the Dart side of the plugin via a method channel?

So I'm guessing I should in the gtk plugin use something like: 
```
static constexpr char kMultilineInputType[] = "TextInputType.multiline";
fl_value_lookup_string(config_value, kMultilineInputType);
```
to get the current value of multiline input type, does that sound right?

**Comment by maks** [comment]
Well its certainly an educational experience setting up to build the engine - I ended up having to do a bit of yak shaving to get it building inside a 18.04 docker container as the Ubuntu 20.04 desktop install I have isn't supported by the build system.

I've had a go at doing a PR for this: https://github.com/flutter/engine/pull/20660

Unfortunately it doesn't completely work at the moment: for some reason the newline char that is sent to the TextField only shows up in the UI after I type a subsequent char - not sure why that's happening?

I'll try to work some more on this tomorrow, if anyone has any suggestions or feedback on what I've done so far its *most* welcome.

**Comment by stuartmorgan** [comment]
> for some reason the newline char that is sent to the TextField only shows up in the UI after I type a subsequent char - not sure why that's happening?

You aren't setting `changed` when updating the model, like the other code in that function, so it's not sending any updates to the framework.

**Comment by maks** [comment]
Thanks @stuartmorgan ! I suspected I had missed something simple like that. That makes it work fine for me now.
I've updated the PR and submitted it now.

And just wanted to say thank you :+1:  as a new comer to the engine, having [the functionality you added to test the locally built engine](https://github.com/flutter/flutter/wiki/The-flutter-tool#using-a-locally-built-engine-with-the-flutter-tool) made it a really nice workflow to quickly build and test small incremental changes I was making and worked really well, even wiht my setup of 2 VSCode windows, one "remotely connected" to a container where I was doing theengine builds and then another one for the Flutter app code I was testing the newly built engine with.
